kong/dns/client.lua

```diff
@@ -0,0 +1,783 @@
+  if err and err:sub(1, 8) == "callback" then
+    log(ALERT, err)
+  end
+
+  log(DEBUG, "cache lookup ", key, " ans:", answers and #answers or "-", " hlv:", hit_level or "-")
+
+  local ctx = ngx.ctx
+  if ctx and ctx.has_timing then
+    req_dyn_hook_run_hooks("timing", "dns:cache_lookup",
+                           (hit_level and hit_level < HIT_L3))
+  end
+
+  -- hit L1 lru or L2 shm
+  if hit_level and hit_level < HIT_L3 then
+    stats_increment(self.stats, key, HIT_LEVEL_TO_NAME[hit_level])
+  end
+
+  if err or answers.errcode then
+    if not err then
+      local src = answers.errcode < CACHE_ONLY_ERROR_CODE and "server" or "client"
+      err = ("dns %s error: %s %s"):format(src, answers.errcode, answers.errstr)
+    end
+
+    table_insert(tries, { name .. ":" .. TYPE_TO_NAME[qtype], err })
+  end
+
+  return answers, err
+end
+
+
+local function get_search_types(self, name, qtype)
+  local input_types = qtype and { qtype } or self.search_types
+  local checked_types = {}
+  local types = {}
+
+  for _, qtype in ipairs(input_types) do
+    if qtype == TYPE_LAST then
+      qtype = get_last_type(self.cache, name)
+    end
+
+    if qtype and not checked_types[qtype] then
+      table_insert(types, qtype)
+      checked_types[qtype] = true
+    end
+  end
+
+  return types
+end
+
+
+local function check_and_get_ip_answers(name)
+  if name:match("^%d+%.%d+%.%d+%.%d+$") then  -- IPv4
+    return {
+      { name = name, class = 1, type = TYPE_A, address = name },
+    }
+  end
+
+  if name:find(":", 1, true) then             -- IPv6
+    return {
+      { name = name, class = 1, type = TYPE_AAAA, address = ipv6_bracket(name) },
+    }
+  end
+
+  return nil
+end
+
+
+-- resolve all `name`s and `type`s combinations and return first usable answers
+local function resolve_names_and_types(self, name, typ, cache_only, short_key, tries, resolved_names)
+  local answers = check_and_get_ip_answers(name)
+  if answers then -- domain name is IP literal
+    answers.ttl = LONG_LASTING_TTL
+    answers.expire = now() + answers.ttl
+    return answers, nil, tries
+  end
+
+  -- TODO: For better performance, it may be necessary to rewrite it as an
+  --       iterative function.
+  local types = get_search_types(self, name, typ)
+  local names = search_names(name, self.resolv, self.hosts)
+
+  local err
+  for _, qtype in ipairs(types) do
+    for _, qname in ipairs(names) do
+      answers, err = resolve_name_type(self, qname, qtype, cache_only,
+                                       short_key, tries, resolved_names)
+      -- severe error occurred
+      if not answers then
+        return nil, err, tries
+      end
+
+      if not answers.errcode then
+        insert_last_type(self.cache, qname, qtype) -- cache TYPE_LAST
+        return answers, nil, tries
+      end
+    end
+  end
+
+  -- not found in the search iteration
+  return nil, err, tries
+end
+
+
+local function resolve_all(self, name, qtype, cache_only, tries, resolved_names)
```

Comment:
Oh I forgot that this might also be called recursively, thus:
```suggestion
local function resolve_all(self, name, qtype, cache_only, tries, resolved_names, has_timing)
```